File: packages/node/src/resources/verifications.ts
import { APITimeoutError } from "../errors.js";
import { Page } from "../pagination.js";
import { performRequest, type ResolvedClientOptions } from "../request.js";
import type { CreateVerificationRequest, VerifyRequest } from "../types/api.js";
import type { Verification, VerificationBundle } from "../types/shared.js";

const TERMINAL_STATUSES = new Set<Verification["status"]>(["completed", "failed"]);

export interface CreateAndWaitOptions {
  /** Total wait budget in ms. Default 5 minutes. */
  timeout?: number;
  /** Initial poll interval in ms. Doubles up to a 30s cap. */
  pollInterval?: number;
  signal?: AbortSignal;
}

export class VerificationsResource {
  constructor(private readonly client: ResolvedClientOptions) {}

  /**
   * Two flavors of "create" exist on the API:
   *
   *   POST /v1/verifications  — requires entity_id; returns a Verification
   *   POST /v1/verify         — looks up/creates the entity then verifies;
   *                             returns a VerificationBundle (sync 200,
   *                             pending 202, ambiguous 409 → ConflictError)
   *
   * `create()` calls `/v1/verify` because it covers both cases. Use
   * `createForEntity()` when you already have an entity id.
   */
  async create(request: VerifyRequest): Promise<VerificationBundle & { _requestId: string | null; _status: number }> {
    const res = await performRequest<VerificationBundle>(this.client, {
      method: "POST",
      path: "/v1/verify",
      body: request,
      autoIdempotency: true,
    });
    return { ...res.data, _requestId: res.requestId, _status: res.status };
  }

  async createForEntity(
    request: CreateVerificationRequest,
  ): Promise<Verification & { _requestId: string | null; _status: number }> {
    const res = await performRequest<Verification>(this.client, {
      method: "POST",
      path: "/v1/verifications",
      body: request,
      autoIdempotency: true,
    });
    return { ...res.data, _requestId: res.requestId, _status: res.status };
  }

  async retrieve(id: string): Promise<Verification & { _requestId: string | null }> {
    const res = await performRequest<Verification>(this.client, {
      method: "GET",
      path: `/v1/verifications/${encodeURIComponent(id)}`,
    });
    return { ...res.data, _requestId: res.requestId };
  }

  async list(query?: { limit?: number; status?: Verification["status"] }): Promise<Page<Verification>> {
    const res = await performRequest<{ data: Verification[] } | Verification[]>(this.client, {
      method: "GET",
      path: "/v1/verifications",
      query: query as Record<string, string | number | boolean | undefined> | undefined,
    });
    const items = Array.isArray(res.data) ? res.data : res.data.data ?? [];
    return new Page(items);
  }

  /**
   * Submit and poll until the verification reaches a terminal status, the
   * timeout elapses, or the signal aborts.
   */
  async createAndWait(
    request: VerifyRequest,
    options: CreateAndWaitOptions = {},
  ): Promise<VerificationBundle> {
    // Honor an already-aborted signal before we make any server-side change.
    throwIfAborted(options.signal);
    const timeoutMs = options.timeout ?? 5 * 60_000;
    const pollMin = options.pollInterval ?? 1_000;
    if (pollMin <= 0) {
      throw new RangeError("pollInterval must be greater than 0");
    }
    const pollMax = 30_000;

    const initial = await this.create(request);
    if (initial.verification.status && TERMINAL_STATUSES.has(initial.verification.status)) {
      return initial;
    }

    const deadline = Date.now() + timeoutMs;
    let interval = pollMin;
    while (Date.now() < deadline) {
      throwIfAborted(options.signal);
      await sleep(Math.min(interval, deadline - Date.now()), options.signal);
      const refreshed = await this.retrieve(initial.verification.id);
      if (TERMINAL_STATUSES.has(refreshed.status)) {
        return {
          object: "verification_bundle",
          entity: initial.entity,
          verification: refreshed,
        };
      }
      interval = Math.min(pollMax, Math.floor(interval * 2));
    }

    throw new APITimeoutError(timeoutMs, initial._requestId);
  }
}

function throwIfAborted(signal: AbortSignal | undefined): void {
  if (signal?.aborted) {
    throw signal.reason instanceof Error ? signal.reason : new Error("Aborted");
  }
}

function sleep(ms: number, signal: AbortSignal | undefined): Promise<void> {
  if (ms <= 0) return Promise.resolve();
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      clearTimeout(timer);
      reject(signal?.reason instanceof Error ? signal.reason : new Error("Aborted"));
    };
    if (signal) {
      if (signal.aborted) {
        clearTimeout(timer);
        reject(signal.reason instanceof Error ? signal.reason : new Error("Aborted"));
        return;
      }
      signal.addEventListener("abort", onAbort, { once: true });
    }
  });
}

File: packages/python/src/vendorval/resources/_verifications.py
"""Verification resource (sync + async) with createAndWait polling."""

from __future__ import annotations

import asyncio
import builtins
import time
from collections.abc import Mapping
from typing import Any

import httpx

from .._errors import APITimeoutError
from .._models import Response, VerificationBundleResponse
from .._pagination import Page
from .._request import ResolvedConfig, execute_async, execute_sync, prepare

_TERMINAL = {"completed", "failed"}


def _build_verify_body(
    *,
    identifiers: list[dict[str, str]],
    checks: list[str],
    legal_name: str | None,
    entity_type: str | None,
    country: str | None,
    address: Mapping[str, Any] | None,
    mode: str | None,
    options: Mapping[str, Any] | None,
) -> dict[str, Any]:
    body: dict[str, Any] = {
        "identifiers": list(identifiers),
        "checks": list(checks),
    }
    if legal_name is not None:
        body["legal_name"] = legal_name
    if entity_type is not None:
        body["entity_type"] = entity_type
    if country is not None:
        body["country"] = country
    if address is not None:
        body["address"] = dict(address)
    if mode is not None:
        body["mode"] = mode
    if options is not None:
        body["options"] = dict(options)
    return body


class VerificationsResource:
    def __init__(self, cfg: ResolvedConfig, client: httpx.Client) -> None:
        self._cfg = cfg
        self._client = client

    def create(
        self,
        *,
        identifiers: list[dict[str, str]],
        checks: list[str],
        legal_name: str | None = None,
        entity_type: str | None = None,
        country: str | None = None,
        address: Mapping[str, Any] | None = None,
        mode: str | None = None,
        options: Mapping[str, Any] | None = None,
    ) -> VerificationBundleResponse:
        body = _build_verify_body(
            identifiers=identifiers,
            checks=checks,
            legal_name=legal_name,
            entity_type=entity_type,
            country=country,
            address=address,
            mode=mode,
            options=options,
        )
        prepared = prepare(
            self._cfg, method="POST", path="/v1/verify", body=body, auto_idempotency=True,
        )
        res = execute_sync(self._client, prepared)
        return VerificationBundleResponse(res.data, res.request_id, res.status)

    def create_for_entity(
        self,
        *,
        entity_id: str,
        checks: list[str],
        mode: str | None = None,
        options: Mapping[str, Any] | None = None,
    ) -> Response:
        body: dict[str, Any] = {"entity_id": entity_id, "checks": list(checks)}
        if mode is not None:
            body["mode"] = mode
        if options is not None:
            body["options"] = dict(options)
        prepared = prepare(
            self._cfg, method="POST", path="/v1/verifications", body=body, auto_idempotency=True,
        )
        res = execute_sync(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

    def retrieve(self, verification_id: str) -> Response:
        prepared = prepare(self._cfg, method="GET", path=f"/v1/verifications/{verification_id}")
        res = execute_sync(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

    def list(
        self,
        *,
        limit: int | None = None,
        status: str | None = None,
    ) -> Page[Response]:
        prepared = prepare(
            self._cfg,
            method="GET",
            path="/v1/verifications",
            query={"limit": limit, "status": status},
        )
        res = execute_sync(self._client, prepared)
        items = res.data if isinstance(res.data, list) else (res.data or {}).get("data", [])
        return Page([Response(item, res.request_id, res.status) for item in items])

    def create_and_wait(
        self,
        *,
        identifiers: builtins.list[dict[str, str]],
        checks: builtins.list[str],
        legal_name: str | None = None,
        entity_type: str | None = None,
        country: str | None = None,
        address: Mapping[str, Any] | None = None,
        mode: str | None = None,
        options: Mapping[str, Any] | None = None,
        timeout: float = 5 * 60.0,
        poll_interval: float = 1.0,
    ) -> VerificationBundleResponse:
        bundle = self.create(
            identifiers=identifiers,
            checks=checks,
            legal_name=legal_name,
            entity_type=entity_type,
            country=country,
            address=address,
            mode=mode,
            options=options,
        )
        verification = bundle.verification
        if verification.get("status") in _TERMINAL:
            return bundle

        deadline = time.monotonic() + timeout
        delay = poll_interval
        verification_id = verification["id"]
        while time.monotonic() < deadline:
            time.sleep(min(delay, max(0.0, deadline - time.monotonic())))
            refreshed = self.retrieve(verification_id)
            if refreshed.get("status") in _TERMINAL:
                return VerificationBundleResponse(
                    {"object": "verification_bundle", "entity": bundle.entity.to_dict(), "verification": refreshed.to_dict()},
                    refreshed.request_id,
                    refreshed.status,
                )
            delay = min(delay * 2, 30.0)

        raise APITimeoutError(timeout, request_id=bundle.request_id)


class AsyncVerificationsResource:
    def __init__(self, cfg: ResolvedConfig, client: httpx.AsyncClient) -> None:
        self._cfg = cfg
        self._client = client

    async def create(
        self,
        *,
        identifiers: list[dict[str, str]],
        checks: list[str],
        legal_name: str | None = None,
        entity_type: str | None = None,
        country: str | None = None,
        address: Mapping[str, Any] | None = None,
        mode: str | None = None,
        options: Mapping[str, Any] | None = None,
    ) -> VerificationBundleResponse:
        body = _build_verify_body(
            identifiers=identifiers,
            checks=checks,
            legal_name=legal_name,
            entity_type=entity_type,
            country=country,
            address=address,
            mode=mode,
            options=options,
        )
        prepared = prepare(
            self._cfg, method="POST", path="/v1/verify", body=body, auto_idempotency=True,
        )
        res = await execute_async(self._client, prepared)
        return VerificationBundleResponse(res.data, res.request_id, res.status)

    async def create_for_entity(
        self,
        *,
        entity_id: str,
        checks: list[str],
        mode: str | None = None,
        options: Mapping[str, Any] | None = None,
    ) -> Response:
        body: dict[str, Any] = {"entity_id": entity_id, "checks": list(checks)}
        if mode is not None:
            body["mode"] = mode
        if options is not None:
            body["options"] = dict(options)
        prepared = prepare(
            self._cfg, method="POST", path="/v1/verifications", body=body, auto_idempotency=True,
        )
        res = await execute_async(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

    async def retrieve(self, verification_id: str) -> Response:
        prepared = prepare(self._cfg, method="GET", path=f"/v1/verifications/{verification_id}")
        res = await execute_async(self._client, prepared)
        return Response(res.data, res.request_id, res.status)

    async def list(
        self,
        *,
        limit: int | None = None,
        status: str | None = None,
    ) -> Page[Response]:
        prepared = prepare(
            self._cfg,
            method="GET",
            path="/v1/verifications",
            query={"limit": limit, "status": status},
        )
        res = await execute_async(self._client, prepared)
        items = res.data if isinstance(res.data, list) else (res.data or {}).get("data", [])
        return Page([Response(item, res.request_id, res.status) for item in items])

    async def create_and_wait(
        self,
        *,
        identifiers: builtins.list[dict[str, str]],
        checks: builtins.list[str],
        legal_name: str | None = None,
        entity_type: str | None = None,
        country: str | None = None,
        address: Mapping[str, Any] | None = None,
        mode: str | None = None,
        options: Mapping[str, Any] | None = None,
        timeout: float = 5 * 60.0,
        poll_interval: float = 1.0,
    ) -> VerificationBundleResponse:
        bundle = await self.create(
            identifiers=identifiers,
            checks=checks,
            legal_name=legal_name,
            entity_type=entity_type,
            country=country,
            address=address,
            mode=mode,
            options=options,
        )
        verification = bundle.verification
        if verification.get("status") in _TERMINAL:
            return bundle

        loop = asyncio.get_event_loop()
        deadline = loop.time() + timeout
        delay = poll_interval
        verification_id = verification["id"]
        while loop.time() < deadline:
            await asyncio.sleep(min(delay, max(0.0, deadline - loop.time())))
            refreshed = await self.retrieve(verification_id)
            if refreshed.get("status") in _TERMINAL:
                return VerificationBundleResponse(
                    {"object": "verification_bundle", "entity": bundle.entity.to_dict(), "verification": refreshed.to_dict()},
                    refreshed.request_id,
                    refreshed.status,
                )
            delay = min(delay * 2, 30.0)

        raise APITimeoutError(timeout, request_id=bundle.request_id)
